tensorflow/core/common_runtime/dml/dml_execution_context.cc

```diff
@@ -372,4 +421,176 @@ void DmlExecutionContextImpl::CloseCommandListAndExecute() {
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::ExecuteOperator(
+    IDMLCompiledOperator* op,
+    Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+    ID3D12DescriptorHeap* descriptor_heap) {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  shared_state_->WriteBatch().emplace_back(
+      [=, binding_table = std::move(binding_table)]() {
+        shared_state_->impl->ExecuteOperator(op, binding_table.Get(),
+                                             descriptor_heap);
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::ResourceBarrier(
+    absl::Span<const D3D12_RESOURCE_BARRIER> barriers) {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  // The caller may not keep the barriers referenced by the span alive for
+  // longer than this function call, so make a copy and transfer ownership to
+  // the lambda.
+  absl::InlinedVector<D3D12_RESOURCE_BARRIER, 4> barriers_copy;
+  shared_state_->WriteBatch().emplace_back(
+      [=, barriers = std::move(barriers_copy)]() {
+        shared_state_->impl->ResourceBarrier(barriers);
+      });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+DmlGpuEvent DmlExecutionContext::UavBarrier() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  shared_state_->WriteBatch().emplace_back(
+      [=]() { shared_state_->impl->UavBarrier(); });
+
+  shared_state_->new_function_enqueued.notify_all();
+
+  return shared_state_->next_flush_event;
+}
+
+StatusOr<DmlGpuEvent> DmlExecutionContext::Flush() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+
+  auto event = shared_state_->next_flush_event;
+  if (shared_state_->WriteBatch().empty()) {
+    --event.fence_value;
+  }
+
+  shared_state_->flush_requested = true;
+  shared_state_->new_function_enqueued.notify_all();
+  return event;
+}
+
+DmlGpuEvent DmlExecutionContext::GetCurrentCompletionEvent() {
+  std::unique_lock<std::mutex> lock(shared_state_->mutex);
+  auto event = shared_state_->next_flush_event;
+
+  if (shared_state_->WriteBatch().empty()) {
+    --event.fence_value;
+  }
+
+  return event;
+}
+
+/*static*/ void DmlExecutionContext::ThreadProc(
+    std::shared_ptr<SharedState> state) {
+  auto last_flush_time = std::chrono::high_resolution_clock::now();
+
+  while (true) {
+    std::chrono::duration<double> elapsed =
+        std::chrono::high_resolution_clock::now() - last_flush_time;
+    auto elapsed_us = elapsed.count() * 1e6;
+
+    std::unique_lock<std::mutex> lock(state->mutex);
+    if (state->exit_requested) {
+      break;
+    }
+
+    auto& batch = state->WriteBatch();
+
+    if (batch.empty()) {
+      // Wait for new work to be batched.
+      state->new_function_enqueued.wait(lock);
+      continue;
```

Comment:
It may be worthwhile to add a comment that this `continue` is necessary to handle spurious wakeups (otherwise to a casual observer it might look superfluous)